xde/src/xde.rs

```diff
@@ -297,17 +307,105 @@ unsafe extern "C" fn _fini() -> c_int {
     }
 }
 
+/// Handle `open(9E)` for non-MAC managed devices.
+///
+/// MAC providers are STREAMS drivers and thus use the `str_ops` entrypoints,
+/// leaving `cb_open` and others typically set to `nodev`. However, the MAC
+/// framework does allow drivers to provide its own set of minor nodes as
+/// regular char/block devices. We create one such device (/dev/xde) in
+/// `xde_attach`. See also `xde_getinfo`.
+/// MAC will return `ENOSTR` from its STREAMS-based `open(9E)` routine if
+/// passed a minor node reserved for driver private use. In that case,
+/// the system will retry the open with the driver's `cb_open` routine.
+#[no_mangle]
+unsafe extern "C" fn xde_open(
+    devp: *mut dev_t,
+    flag: c_int,
+    otyp: c_int,
+    credp: *mut cred_t,
+) -> c_int {
+    assert!(!xde_dip.is_null());
+
+    if otyp != OTYP_CHR {
+        return EINVAL;
+    }
+
+    let minor = getminor(*devp);
+    if minor != XDE_CTL_MINOR {
+        return ENXIO;
+    }
+
+    match secpolicy::secpolicy_dl_config(credp) {
+        0 => {}
+        err => {
+            warn!("secpolicy_dl_config failed: {err}");
+            return err;
+        }
+    }
+
+    let flag = flag & !(FCLOEXEC | FOFFMAX);
```

Comment:
Is there a reason you had to mask those flags off? They've never come up for me in the driver work. In particular, no driver today checks `FCLOEXEC` or `FOFFMAX`. Both of these are noted as open-only so I don't think they make it to a driver (along with many others).

Reply:
Those are the flags I'm seeing:

```
$ pfexec dtrace -n 'xde_open:entry { stack(); ustack(); printf("flags = 0x%x", arg1);  }' -c './target/release/opteadm list-ports' | rustfilt 
dtrace: description 'xde_open:entry ' matched 1 probe
dtrace: pid 633125 has exited
LINK                             MAC ADDRESS              IPv4 ADDRESS     EXTERNAL IPv4    IPv6 ADDRESS                             EXTERNAL IPv6                            STATE   
CPU     ID                    FUNCTION:NAME
  4  82780                   xde_open:entry 
              genunix`dev_open+0x26
              specfs`spec_open+0x18e
              genunix`fop_open+0xa4
              genunix`vn_openat+0x208
              genunix`copen+0x431
              genunix`openat+0x29
              genunix`open+0x1c
              unix`sys_syscall+0x283

              libc.so.1`_syscall6+0x1b
              libc.so.1`open+0xaf
              opteadm`std::sys::unix::fs::File::open_c+0xf1
              opteadm`std::fs::OpenOptions::_open+0x88
              opteadm`opteadm::OpteAdm::open+0x50
              opteadm`opteadm::main+0x15fb
              opteadm`std::sys_common::backtrace::__rust_begin_short_backtrace+0x6
              opteadm`std::rt::lang_start::{{closure}}+0x11
              opteadm`std::rt::lang_start_internal+0x2c6
              opteadm`main+0x2c
              opteadm`_start_crt+0x87
              opteadm`_start+0x18
flags = 0x802003
```

`0x802003` is `FCLOEXEC | FOFFMAX | FWRITE | FREAD`

Reply:
Hmm, ok. I guess the reality is that drivers aren't supposed to worry about those. I think I gave bad advise then in terms of phrasing it above and we should look for explicit flags that we don't support and error on those.  Looking at how I've phrased this in recent drivers I've done:

```
if ((flags & (FEXCL | FNDELAY | FNONBLOCK)) != 0) {
```

Given that other open flags can get added over time let's just look for the specific ones. I don't think we really want to be looking at `FCLOEXEC`, `FOFFMAX`, etc. This really is the only driver that would be. Sorry about making a mess of this one.

Reply:
Thanks, looks good.